Call #prerecord when running tests in parallel with processes

ricardotk002 opened the pull request:
### Summary

Fixes #41766

For the sake of consistency with what `Minitest` does for serial and
parallel tests with threads, some reportes may rely on this:

https://github.com/seattlerb/minitest/blob/3c6576a/lib/minitest.rb#L339-L342
https://github.com/seattlerb/minitest/blob/3c6576a/lib/minitest/parallel.rb#L32-L34

The steps are:

* Prerecord
* Run test
* Record

When running in parallel, some `prerecord` calls may arrive later than
their `record` counterparts, e.g.:

* prerecordA
* prerecordB
* recordB
* recordA

### Other Information

I fixed an error in the reporter tests file, it fails when is executed using `bin/test`:

```
railties $ bin/test test/test_unit/reporter_test.rb
```

ricardotk002 committed:
Call #prerecord when running tests in parallel with processes

For the sake of consistency with what `Minitest` does for serial and
parallel tests with threads, some reportes may rely on this:

https://github.com/seattlerb/minitest/blob/3c6576a/lib/minitest.rb#L339-L342
https://github.com/seattlerb/minitest/blob/3c6576a/lib/minitest/parallel.rb#L32-L34

The steps are:

* Prerecord
* Run test
* Record

When running in parallel, some `prerecord` calls may arrive later than
their `record` counterparts, e.g.:

* prerecordA
* prerecordB
* recordB
* recordA

ricardotk002 committed:
Fix reporter test that fails when using the executable bin/test

This particular test fails when is executed using `bin/test`

```
railties $ bin/test test/test_unit/reporter_test.rb
```

Because the value in `Rails::TestUnitReporter.executable` is leaking from the test
setup to the tests themselves. It works fine with `rake test` or
`rails test`.

This could also be done with the following, but you may get misleading
info about your own tests.

```diff
diff --git a/railties/test/test_unit/reporter_test.rb b/railties/test/test_unit/reporter_test.rb
index d9140648fa..ef3ab976b1 100644
--- a/railties/test/test_unit/reporter_test.rb
+++ b/railties/test/test_unit/reporter_test.rb
@@ -12,6 +12,12 @@ def woot; end
   setup do
     @output = StringIO.new
     @reporter = Rails::TestUnitReporter.new @output, output_inline: true
+    @old_executable = Rails::TestUnitReporter.executable
+    Rails::TestUnitReporter.executable = "rails test"
+  end
+
+  teardown do
+    Rails::TestUnitReporter.executable = @old_executable
   end

   test "prints rerun snippet to run a single failed test" do
```

hurricup commented:
As I can see in the code, `Minitest` itself does not do something about tests overlapping with thread parallelism, it runs exactly pre-record (under lock), test in thread and record (under lock). 
I believe AS done the same. 

Reporters should properly such overlapping test results and I believe they do, otherwise thread-based parallelism couldn't work properly.

eileencodes commented:
@hurricup the only thread-based parallelism for tests in active support directly implements Minitest. It's just a [wrapper](https://github.com/rails/rails/blob/ad9e52066cde3a7e5a8c1171206c11d6922f2068/activesupport/lib/active_support/test_case.rb#L84). 

The default parallelism that is implemented with Drb uses processes and not threads. That implementation is what is changing in this PR.

hurricup commented:
@eileencodes I'm not sure I get what are you trying to say.
My point that suggested approach differs from `Minitest`'s for unclear reason. 
Minitest has it's own thread-based parallelism and does not make simultaneous prerecord/record invocations.

eileencodes commented:
I'm saying that this PR has nothing to do with threads because the parallelizer that Rails implements is process-based.

hurricup commented:
Probably we have some kind of miscommunication because of my poor English, sorry.

ricardotk002 commented:
I tried to follow the Minitest's approach: calling pre-record from the worker object, right before running the tests. However, since both steps (prerecord, record) can't be synced between forks, the calls could arrive in a different order, leading to wrong reports. Probably this can be done if we switch from a queue to a pipe.

Given that prerecord only notifies about the class and method name that is about to run, I thought it made sense to send them in order at the same block. Unfortunately, this will could give you inexact results if you're relying on it for timing events.

@eileencodes I think we can close this one, I'm not experienced enough with Drb to properly support the Minitest interface. Do you think this commit 44e4d72 still makes sense? I can send it in another PR.

hurricup commented:
@ricardotk002 sorry, I still don't understand what is wrong with Minitest approach. 

As far as I get it in `rails`: single worker is responsible for processing a test. So worker first invokes `prerecord` via drb, then runs a test, then invokes `record`.

Can't imagine how `record` may occur before `prerecord` with such approach and which forks need to be synchronized.

Yes, we can have situation when server gets:

1. prerecordA
2. prerecordB
3. recordB
4. recordA

And this situation is totally fine, and may happen in minitest itself when Minitest's thread based parallelization is used. 

And as far as I get it we will never get:

1. recordA
2. prerecordA

Or I don't understand something.

ricardotk002 commented:
@hurricup Yes, that's the case I was trying to explain. You're correct in that it also happens in threaded parallelization. I changed the code so `prerecord` now runs before the test execution. I also reverted my changes to `Rails::TestUnitReporter` as it could cause broken reports when using the verbose option.

donv commented:
Would be nice to get this merged.  Test reporting of parallel tests in RubyMine is broken until then.  Inserting the changes in this PR in my `test/test_helper.rb` works as a workaround for me:

```ruby
class ActiveSupport::Testing::Parallelization::Worker
  def perform_job(job)
    klass = job[0]
    method = job[1]
    reporter = job[2]

    set_process_title("#{klass}##{method}")

    @queue.prerecord(reporter, klass, method)

    result = klass.with_info_handler reporter do
      Minitest.run_one_method(klass, method)
    end
    safe_record(reporter, result)
  end
end

class ActiveSupport::Testing::Parallelization::Server
  def prerecord(reporter, klass, method)
    reporter.synchronize do
      reporter.prerecord(klass, method)
    end
  end
end
```

Thank you all for your great work!

ricardotk002 commented:
@eileencodes I rebased it with the latest version in `main`. I tested it with RubyMine 2021.2 but didn't notice a difference really. As discussed previously, it seems like a good idea to mimic the `Minitest` API as much as possible. The only failure in the CI seems unrelated. Let me know if there's something else I can help with here.

eileencodes commented:
> I tested it with RubyMine 2021.2 but didn't notice a difference really

I'm confused - does that mean this PR doesn't change anything?

hurricup commented:
The difference noticeble on long running tests. When test started event is not received by IDE until the test is finished.

ricardotk002 commented:
> I'm confused - does that mean this PR doesn't fix anything?

Actually the issue is more noticeable with long running tests as @hurricup describes. Notice how the tests aren't announced (loading) by `#prerecord` in the first capture, compared with the second one.

Before patch:
![before](https://user-images.githubusercontent.com/3733470/127889659-f183c3a4-521d-4725-ad69-2708d169788f.gif)

After patch (tests 1~5 were delayed with a `#sleep` call)
![after](https://user-images.githubusercontent.com/3733470/127889667-396f976d-3555-4bda-ad4b-d9cebaf126c1.gif)

donv commented:
This is still a problem.
